torch/csrc/distributed/c10d/ProcessGroupXCCL.cpp

```diff
@@ -0,0 +1,259 @@
+#include <torch/csrc/distributed/c10d/ProcessGroupXCCL.hpp>
+#include <fstream>
+#include <mutex>
+#include <sstream>
+
+#ifdef USE_C10D_XCCL
+#include <exception>
+#include <map>
+#include <stdexcept>
+#include <tuple>
+#include <unordered_set>
+#include <utility>
+
+#include <c10/core/DeviceType.h>
+#include <c10/util/CallOnce.h>
+#include <c10/util/Exception.h>
+#include <c10/util/Logging.h>
+#include <c10/util/Optional.h>
+#include <c10/util/irange.h>
+#include <torch/csrc/distributed/c10d/ParamCommsUtils.hpp>
+#include <torch/csrc/distributed/c10d/TraceUtils.h>
+#include <torch/csrc/distributed/c10d/Utils.hpp>
+#include <torch/torch.h>
+
+namespace c10d {
+
+namespace {
+std::map<c10d::ReduceOp, ccl::reduction> xcclOps = {
+    {ReduceOp::MIN, ccl::reduction::min},
+    {ReduceOp::MAX, ccl::reduction::max},
+    {ReduceOp::SUM, ccl::reduction::sum},
+    {ReduceOp::PRODUCT, ccl::reduction::prod},
+};
+
+std::map<at::ScalarType, ccl::datatype> xcclDatatypes = {
+    {at::kByte, ccl::datatype::uint8},
+    {at::kChar, ccl::datatype::int8},
+    {at::kShort, ccl::datatype::int16},
+    {at::kInt, ccl::datatype::int32},
+    {at::kLong, ccl::datatype::int64},
+    {at::kHalf, ccl::datatype::float16},
+    {at::kFloat, ccl::datatype::float32},
+    {at::kDouble, ccl::datatype::float64},
+    {at::kBFloat16, ccl::datatype::bfloat16},
+    {at::kBool, ccl::datatype::uint8},
+};
+
+void check_gpu_single_tensor(const at::Tensor& tensor) {
```

Comment:
Hi, is it better to be `check_xpu_single_tensor` since it require tensor on xpu? :D